crates/accelerate/src/sparse_observable.rs

```diff
@@ -0,0 +1,1677 @@
+    /// error condition.
+    #[inline]
+    fn try_from_u8(value: u8) -> Result<Option<Self>, BitTermFromU8Error> {
+        match value {
+            b'+' => Ok(Some(BitTerm::Plus)),
+            b'-' => Ok(Some(BitTerm::Minus)),
+            b'0' => Ok(Some(BitTerm::Zero)),
+            b'1' => Ok(Some(BitTerm::One)),
+            b'I' => Ok(None),
+            b'X' => Ok(Some(BitTerm::X)),
+            b'Y' => Ok(Some(BitTerm::Y)),
+            b'Z' => Ok(Some(BitTerm::Z)),
+            b'l' => Ok(Some(BitTerm::Left)),
+            b'r' => Ok(Some(BitTerm::Right)),
+            _ => Err(BitTermFromU8Error(value)),
+        }
+    }
+}
+
+static BIT_TERM_PY_ENUM: GILOnceCell<Py<PyType>> = GILOnceCell::new();
+static BIT_TERM_INTO_PY: GILOnceCell<[Option<Py<PyAny>>; 16]> = GILOnceCell::new();
+
+/// Construct the Python-space `IntEnum` that represents the same values as the Rust-spce `BitTerm`.
+///
+/// We don't make `BitTerm` a direct `pyclass` because we want the behaviour of `IntEnum`, which
+/// specifically also makes its variants subclasses of the Python `int` type; we use a type-safe
+/// enum in Rust, but from Python space we expect people to (carefully) deal with the raw ints in
+/// Numpy arrays for efficiency.
+///
+/// The resulting class is attached to `SparseObservable` as a class attribute, and its
+/// `__qualname__` is set to reflect this.
+fn make_py_bit_term(py: Python) -> PyResult<Py<PyType>> {
+    let terms = [
+        BitTerm::X,
+        BitTerm::Plus,
+        BitTerm::Minus,
+        BitTerm::Y,
+        BitTerm::Right,
+        BitTerm::Left,
+        BitTerm::Z,
+        BitTerm::Zero,
+        BitTerm::One,
+    ]
+    .into_iter()
+    .flat_map(|term| {
+        let mut out = vec![(term.py_name(), term as u8)];
+        if term.py_name() != term.py_label() {
+            // Also ensure that the labels are created as aliases.  These can't be (easily) accessed
+            // by attribute-getter (dot) syntax, but will work with the item-getter (square-bracket)
+            // syntax, or programmatically with `getattr`.
+            out.push((term.py_label(), term as u8));
+        }
+        out
+    })
+    .collect::<Vec<_>>();
+    let obj = py.import_bound("enum")?.getattr("IntEnum")?.call(
+        ("BitTerm", terms),
+        Some(
+            &[
+                ("module", "qiskit.quantum_info"),
+                ("qualname", "SparseObservable.BitTerm"),
+            ]
+            .into_py_dict_bound(py),
+        ),
+    )?;
+    Ok(obj.downcast_into::<PyType>()?.unbind())
+}
+
+// Return the relevant value from the Python-space sister enumeration.  These are Python-space
+// singletons and subclasses of Python `int`.  We only use this for interaction with "high level"
+// Python space; the efficient Numpy-like array paths use `u8` directly so Numpy can act on it
+// efficiently.
+impl IntoPy<Py<PyAny>> for BitTerm {
```

Comment:
The reason to have a bit of a split is that PyO3's `pyclass` derivations for Rust `enum`s don't work as pleasantly in Python as proper Python enumerations.  For example, making the Python-exposed class an `IntEnum` means that all its elements are subclasses of `int` properly, so you can put them in a Numpy array of `u8`, which means you can use the proper named access if you like, rather than having to deal with the magic numbers.

Reply:
Ohh I see, right, so PyO3 `pyclass`es can pretty much only be standalone classes but not derive from a Python type like `enum.IntEnum`?

Reply:
There are mechanisms to make them subclass each other, but I'm not sure about types defined from Python space (like `IntEnum`). Either way, they're really really fiddly to work with if you do it.